airflow/providers/amazon/aws/fs/s3.py

```diff
@@ -0,0 +1,123 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+import logging
+from functools import partial
+from typing import TYPE_CHECKING, Any, Callable, Dict
+
+import requests
+from botocore import UNSIGNED
+from requests import HTTPError
+
+from airflow.providers.amazon.aws.hooks.base_aws import AwsGenericHook
+
+if TYPE_CHECKING:
+    from botocore.awsrequest import AWSRequest
+    from fsspec import AbstractFileSystem
+
+
+Properties = Dict[str, str]
+
+S3_PROXY_URI = "proxy-uri"
+
+log = logging.getLogger(__name__)
+
+schemes = ["s3", "s3a", "s3n"]
+
+
+class SignError(Exception):
+    """Raises when unable to sign a S3 request."""
+
+
+def get_fs(conn_id: str | None) -> AbstractFileSystem:
+    try:
+        from s3fs import S3FileSystem
+    except ImportError:
+        raise ImportError(
+            "Airflow FS S3 protocol requires the s3fs library, but it is not installed as it requires"
+            "aiobotocore. Please install the s3 protocol support library by running: "
+            "pip install apache-airflow[s3]"
+        )
+
+    aws: AwsGenericHook = AwsGenericHook(aws_conn_id=conn_id, client_type="s3")
+    session = aws.get_session(deferrable=True)
+    endpoint_url = aws.conn_config.extra_config.get("endpoint_url", None)
```

Comment:
```suggestion
    endpoint_url = aws.conn_config.get_service_endpoint_url(service_name="s3")
```

This would resolve [AWS Service Endpoint URL configuration](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/connections/aws.html#aws-service-endpoint-url-configuration)